prboom2/src/p_maputl.c

```diff
@@ -753,23 +752,28 @@ static mobj_t *RoughBlockCheck(mobj_t *mo, int index, angle_t fov)
       continue;
     }
 
-    // skip other players in co-op
-    if (netgame && !deathmatch && link->player)
+    // skip the projectile's owner
+    if (link == mo->target)
     {
       link = link->bnext;
       continue;
     }
 
-    // skip the projectile's owner
-    if (link == mo->target)
+    // skip friendly actors if owner is friendly,
+    // except for other players during deathmatch.
+    if (mo->target &&
+      (link->flags & mo->target->flags & MF_FRIEND) &&
+      (!deathmatch || !link->player))
     {
       link = link->bnext;
       continue;
     }
 
-    // skip friendly actors if owner is friendly (or a player)
-    if ((link->flags & MF_FRIEND) && mo->target != NULL &&
-      ((mo->target->flags & MF_FRIEND) || mo->target->player))
+    // likewise, skip non-friendly actors if owner is non-friendly
+    // (unless explicitly targeting another non-friendly, so infighting works)
+    if (mo->target &&
+      !((link->flags ^ mo->target->flags) & MF_FRIEND) &&
```

Comment:
This check is redundant with the one above (the xor accounts for all combinations). How about removing the separate one above and just adding this:
`&& !(deathmatch && link->player && mo->target->player)`

Other reason: friend monsters can infight, right? Then the friend / deathmatch condition above will skip friends who are infighting despite the second condition aiming to not skip this case.

Reply:
Sorry, all these rules make my head spin. It's probably better if you make the edits on this one.